src/lib/storage/lz4_segment.cpp

```diff
@@ -13,8 +13,8 @@ template <typename T>
 LZ4Segment<T>::LZ4Segment(const pmr_vector<char>& compressed_data, const pmr_vector<bool>& null_values,
                           const std::shared_ptr<const pmr_vector<size_t>>& offsets, const size_t decompressed_size)
     : BaseEncodedSegment{data_type_from_type<T>()},
-      _compressed_data{std::move(compressed_data)},
-      _null_values{std::move(null_values)},
+      _compressed_data{compressed_data},
+      _null_values{null_values},
```

Comment:
Maybe change the signature to accept by universal reference; that way you can actually move the data instead of copying it.

Reply:
Hm, wouldn't changing the parameters to non-const make them movable? I can do that (and that is what happens in other encoded segments as well).

Reply:
http://blogs.microsoft.co.il/sasha/2014/08/21/c-sink-parameter-passing/

Making it a regular reference would consume it without letting the caller know that you do so. Example (from [here](https://stackoverflow.com/questions/45514116/have-rvalue-reference-instead-of-forwarding-reference-with-variadic-template)):

```
auto foo(X&& arg) -> void {...};

X x;
foo(x);            // compile error [OK]
foo(std::move(x)); // accepted      [OK]
foo(X{});          // accepted      [OK]
```

This is good because you don't inadvertently destroy x.

Reply:
I can't quite get that working (maybe since the move is in the constructor?). The final solution(s) of that blog post seem kind of overkill and the segment should only ever be created by the encoder (which has no other job and does not use the resulting encoded segment).

Since the other implementations of encoded segments either do the same (move a non-const value) or use shared pointers, I would propose to keep it as it is and maybe create an issue for it (covering all encoded segments).

Reply:
Not sure what the problem is:

```
diff --git a/src/lib/storage/lz4_segment.cpp b/src/lib/storage/lz4_segment.cpp
index e42d60926..d444e1454 100644
--- a/src/lib/storage/lz4_segment.cpp
+++ b/src/lib/storage/lz4_segment.cpp
@@ -10,7 +10,7 @@
 namespace opossum {

 template <typename T>
-LZ4Segment<T>::LZ4Segment(pmr_vector<char> compressed_data, pmr_vector<bool> null_values,
+LZ4Segment<T>::LZ4Segment(pmr_vector<char>&& compressed_data, pmr_vector<bool>&& null_values,
                           const std::shared_ptr<const pmr_vector<size_t>>& offsets, const size_t decompressed_size)
     : BaseEncodedSegment{data_type_from_type<T>()},
       _compressed_data{std::move(compressed_data)},
@@ -121,7 +121,7 @@ std::shared_ptr<BaseSegment> LZ4Segment<T>::copy_using_allocator(const Polymorph
     new_offsets_ptr = std::allocate_shared<pmr_vector<size_t>>(alloc, *_offsets);
   }

-  return std::allocate_shared<LZ4Segment>(alloc, new_compressed_data, new_null_values, new_offsets_ptr,
+  return std::allocate_shared<LZ4Segment>(alloc, std::move(new_compressed_data), std::move(new_null_values), new_offsets_ptr,
                                           _decompressed_size);
 }

diff --git a/src/lib/storage/lz4_segment.hpp b/src/lib/storage/lz4_segment.hpp
index 0eb978fc5..a509c6d1d 100644
--- a/src/lib/storage/lz4_segment.hpp
+++ b/src/lib/storage/lz4_segment.hpp
@@ -35,7 +35,7 @@ class LZ4Segment : public BaseEncodedSegment {
    * @param compressed_size The size of the compressed data vector (the return value of LZ4)
    * @param decompressed_size The size in bytes of the decompressed data vector.
    */
-  explicit LZ4Segment(pmr_vector<char> compressed_data, pmr_vector<bool> null_values,
+  explicit LZ4Segment(pmr_vector<char>&& compressed_data, pmr_vector<bool>&& null_values,
                       const std::shared_ptr<const pmr_vector<size_t>>& offsets, const size_t decompressed_size);

   const pmr_vector<bool>& null_values() const;
```

Reply:
Oh ic. I forgot to change it in `copy_using_allocator` as well 😅 .